mimic/model/identity_objects.py

```diff
@@ -0,0 +1,305 @@
+        """
+        if tenant_id in self.endpoints_for_tenants:
+            if template_id in self.endpoints_for_tenants[tenant_id]:
+                self.endpoints_for_tenants[tenant_id].remove(template_id)
+                return
+
+        # Tell the caller if the template did not exist in case the caller
+        # needs to generate error messages
+        raise ValueError(
+            "template (" + template_id + ") not enabled for tenant id ("
+            + tenant_id + ")"
+        )
+
+    def list_templates(self):
+        """
+        List the available templates.
+
+        :returns: an iterable of the end-point templates
+        """
+        return self.endpoint_templates.values()
+
+    def add_template(self, endpoint_template):
+        """
+        Add a new template for the external API.
+
+        :param unicode templates: a :obj:`IEndPointTemplate` to add to the
+            :obj:`IExternalAPIMock` instance
+        :raises: ValueError if the end-point template has already been added
+        :raises: TypeError if the end-point template does not implement
+            the expected interface (IEndPointTempalte)
+        """
+        if IEndPointTemplate.providedBy(endpoint_template):
+            key = endpoint_template.id_key
+
+            if key in self.endpoint_templates:
+                raise ValueError(key + " already exists. Please call update.")
+
+            self.endpoint_templates[key] = endpoint_template
+        else:
+            raise TypeError(
+                endpoint_template.__class__.__module__ + "/" +
+                endpoint_template.__class__.__name__ +
+                " does not implement IEndPointTemplate"
+            )
+
+    def update_template(self, endpoint_template):
+        """
+        Update an existing template for the external API.
+
+        :param unicode templates: a :obj:`IEndPointTemplate` to add to the
+            :obj:`IExternalAPIMock` instance
+        :raises: IndexError if the end-point template does not already exist
+        :raises: TypeError if the end-point template does not implement
+            the expected interface (IEndPointTempalte)
+        """
+        if IEndPointTemplate.providedBy(endpoint_template):
+            key = endpoint_template.id_key
+
+            if key not in self.endpoint_templates:
+                raise IndexError(
+                    "End-point template does not exist. Unable to update. The "
+                    "template must first be added before it can be updated"
+                )
+
+            self.endpoint_templates[key] = endpoint_template
+
+        else:
+            raise TypeError(
+                endpoint_template.__class__.__module__ + "/" +
+                endpoint_template.__class__.__name__ +
+                " does not implement IEndPointTemplate"
+            )
+
+    def remove_template(self, template_id):
+        """
+        Add a new template for the external API.
+
```

Comment:
I believe this docstring should read `Remove a templated endpoint from the external API` or something similar.